File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.java
import java.io.*;
import java.util.*;

public class Main {
	
	private static BufferedReader br;
	private static BufferedWriter bw;
	private static StringTokenizer st;
	private static List<Edge>[] graph;
	private static boolean[] visited;
	private static long answer;
	
	static class Edge implements Comparable<Edge> {
		
		int node;
		int weight;
		
		public Edge(int node, int weight) {
			
			this.node = node;
			this.weight = weight;
		}
		
		@Override
		public int compareTo(Edge e) {
			return this.weight - e.weight;
		}
	}
	
	public static void main(String[] args) throws IOException {
		
		br = new BufferedReader(new InputStreamReader(System.in));
		bw = new BufferedWriter(new OutputStreamWriter(System.out));
		
		st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken()); 
		int M = Integer.parseInt(st.nextToken());
		
		graph = new ArrayList[N + 1];
		visited = new boolean[N + 1];
		answer = 0L;
		
		for (int i = 1; i <= N; i++) {
			graph[i] = new ArrayList<>();
		}
		
		
		for (int i = 0; i < M; i++) {
			
			st = new StringTokenizer(br.readLine());
			int start = Integer.parseInt(st.nextToken());
			int end = Integer.parseInt(st.nextToken());
			int weight = Integer.parseInt(st.nextToken());
			
			graph[start].add(new Edge(end, weight));
			graph[end].add(new Edge(start, weight));
		}
		
		prim();
		
		bw.write(String.valueOf(answer));
		bw.flush();
		bw.close();
		
	}
	
	private static void prim() {
		PriorityQueue<Edge> queue = new PriorityQueue<>();
		queue.offer(new Edge(1,0));
		
		while (!queue.isEmpty()) {
			
			Edge edge = queue.poll();
			int node = edge.node;
			
			if (visited[node]) {
				continue;
			}
			answer += edge.weight;
			
			visited[node] = true;
			
			for (Edge e : graph[node]) {
				if (visited[e.node]) {
					continue;
				}
				
				queue.offer(new Edge(e.node, e.weight));
			}
		}
		
	}

}

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./example.txt";
const inputs = fs.readFileSync(filePath).toString().trim().split("\n");

const [V, E] = inputs.shift().split(" ").map(Number);
let graph = Array.from({length : V + 1}, () => new Array());
const visited = new Array(V + 1).fill(false);

class Edge {

    constructor(node, weight) {
        this.node = node;
        this.weight = weight;
    }

}

class PriorityQueue {

    constructor() {
        this.queue = [0];
    }

    compare(e1, e2) {
        return e1.weight - e2.weight;
    }

    isEmpty() {
        return this.queue.length <= 1;
    }

    offer(edge) {

        let idx = this.queue.length;

        while(idx > 1 && this.compare(edge, this.queue[Math.floor(idx / 2)]) <= 0) {
            this.queue[idx]  = this.queue[Math.floor(idx / 2)];
            idx = Math.floor(idx / 2);
        }

        this.queue[idx] = edge;

    }

    poll() {
        if (this.queue.length <= 1) {
            throw new Error("NoSuchElementException");
        }

        let first = this.queue[1];
        let last = this.queue.pop();
        let size = this.queue.length - 1;

        if (size === 0) {
            return first;
        }

        let pIdx = 1;
        let cIdx = 2;

        while (cIdx <= size) {

            if (this.queue[cIdx + 1] !== undefined && this.compare(this.queue[cIdx], this.queue[cIdx + 1]) > 0) {
                cIdx++;
            }

            if (this.compare(last, this.queue[cIdx]) <= 0) {
                break;
            }

            this.queue[pIdx] = this.queue[cIdx];

            pIdx = cIdx;
            cIdx *= 2;
        }


        this.queue[pIdx] = last;

        return first;
    }

    peek() {
        return this.queue[1];
    }

    size() {
        return this.queue.length - 1;
    }

    clear() {
        this.queue = [0];
    }

}

for (let input of inputs) {
    let [start, end, weight] = input.split(" ").map(Number);
    graph[start].push(new Edge(end, weight));
    graph[end].push(new Edge(start, weight));
}

main();

function main() {
    console.log(prim());
}

function prim() {

    let answer = 0;
    let queue = new PriorityQueue();
    queue.offer(new Edge(1,0));

    while (!queue.isEmpty()) {

        let edge = queue.poll();
        let node = edge.node;

        if (visited[node]) {
            continue;
        }
        visited[node] = true;
        answer += edge.weight;

        for (let e of graph[node]) {
            if (!visited[e.node]) {
                queue.offer(new Edge(e.node, e.weight));
            }
        }

    }

    return answer;

}
